### PR TITLE
Streaming Dataset: Resolve chunks eviction

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR fixes a regression with the StreamingDataset where the chunks won't be deleted as we continue training.

The core reason was:

This would lead to download `max_pre_download + 1` chunks

```python
            if self._pre_download_counter <= self._max_pre_download:
                chunk_index = _get_from_queue(self._to_download_queue)
                if chunk_index == _END_TOKEN:
                    return
```

and the condition to delete was too strict, causing it to be false always.

```python
    def _can_delete_chunk(self) -> bool:
        if self._delete_chunks_when_processed:
            return self._pre_download_counter == self._max_pre_download - 1
```

This PR loosen the condition for deletion but also prevents downloading the extra chunk.

Using the lightning.ai platform with this script and this PR:

```python
import os
from torch.utils.data import DataLoader
from lightning_cloud.utils import add_s3_connection
from lightning.data import StreamingDataset
from lightning.data.streaming.item_loader import TokensLoader
from lit_gpt.packed_dataset import CombinedDataset
from tqdm import tqdm

# Add an external S3 bucket containing some data (StarCoder)
add_s3_connection("tinyllama-template")

# Increase by one because we need the next word as well
effective_block_size = 2048 + 1

input_dir = "/teamspace/s3_connections/tinyllama-template"
train_datasets = [
    StreamingDataset(
        input_dir=f"{input_dir}/slimpajama/train",
        item_loader=TokensLoader(block_size=effective_block_size),
        shuffle=True,
        drop_last=True,
    ),
    StreamingDataset(
        input_dir=f"{input_dir}/starcoder/",
        item_loader=TokensLoader(block_size=effective_block_size),
        shuffle=True,
        drop_last=True,
    ),
]

# Mix SlimPajama data and Starcoder data with these proportions:
weights = (0.693584, 0.306416)
combined_dataset = CombinedDataset(datasets=train_datasets, seed=42, weights=weights)
train_dataloader = DataLoader(combined_dataset, batch_size=8, pin_memory=True, num_workers=os.cpu_count())

# Iterate over the train datasets
for batch in tqdm(train_dataloader):
    pass
```

The disk size remains constant.

Screenshot 2023-12-27 at 09 36 29

Also, this test would fail on master due to the extra chunk:

```python
        thread.download([0, 1, 2, 3, 4, 5, _END_TOKEN])
    
        sleep(0.25)
    
>       assert thread._pre_download_counter == 2
E       assert 3 == 2
E        +  where 3 = <PrepareChunksThread(Thread-2, started daemon 6292156416)>._pre_download_counter

tests/tests_data/streaming/test_reader.py:118: AssertionError
``` 

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19214.org.readthedocs.build/en/19214/

### CODE REVIEW

**Review by andyland — APPROVED** [review]

**Review by lantiga — APPROVED** [review]
Good catch

File: tests/tests_data/streaming/test_reader.py [hunk 1]
```diff
@@ -95,8 +70,34 @@ def test_prepare_chunks_thread(tmpdir):
...
+    thread = PrepareChunksThread(cache._reader.config, item_loader=PyTreeLoader(), max_cache_size=1)
+    assert thread._delete_chunks_when_processed
+
+    thread.start()
+
+    assert thread._pre_download_counter == 0
+
+    thread.download([0, 1, 2, 3, 4, 5, _END_TOKEN])
+
+    sleep(0.25)
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_reader.py:87]: These sleeps can be a source of flakiness. Can you replace them with a barrier that ensures we have downloaded before moving on? We should not rely on this sleep if the task needs to complete in a thread.

For example, you could wait here until `thread._pre_download_counter > 0`, and then run the assertions.

File: tests/tests_data/streaming/test_reader.py [hunk 2]
```diff
@@ -95,8 +70,34 @@ def test_prepare_chunks_thread(tmpdir):
...
+
+    assert thread._pre_download_counter == 2
+    assert not thread._has_exited
+
+    for i in range(5):
+        assert thread._pre_download_counter <= 2
+        thread.delete([i])
+        sleep(0.25)
```
Comment (awaelchli) [inline tests/tests_data/streaming/test_reader.py:95]: same comment here

File: tests/tests_data/streaming/test_reader.py (hunk 1 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_reader.py]: done

File: tests/tests_data/streaming/test_reader.py (hunk 2 above)
Comment (tchaton) [reply tests/tests_data/streaming/test_reader.py]: done

**Review by awaelchli — APPROVED** [review]
